### DISCUSSION TITLE
Claims as Type Parameters

### DISCUSSION THREAD

**Discussion by costela** [golang-jwt/jwt#269] [General]
Hi,

I couldn't find this subject being discussed anywhere else. Sorry for the noise if I missed it!

Is there a consensus opinion on adding type parameter ("generics") support for v5? I'd like to be able to do something like the following without multiple type assertions:
```go
type Foo struct {
  Bar int
  jwt.RegisteredClaims
}

token, err := jwt.ParseWithClaims("ey...", func(token *jwt.Token[*Foo]) (interface{}, error) {
    return calculateKeyFromSomeClaim(token.Claims.Bar) // no type assertion 🎉
})

doSomethingWith(token.Claims.Bar) // no type assertion 🎉
```

My first naive approach working on the v5 branch doesn't feel very clean, so I wanted to gather some input putting more time into it.

Maybe if there are already plans to deprecate some options or change some more APIs for v5, this could become cleaner?

WDYT?

**Comment by oxisto** [comment]
I was playing around with the idea; maybe my branch is even still left over somewhere. The problem I remember was that we need to type assert a function (method) within a non-type asserted struct. And that is not possible. I don't remember exactly which struct it was, I think it was the Parser. Personally I want to release v5 "soon" and maybe focus on this in v6?

**Comment by oxisto** [comment]
So I had another round of experiments (https://github.com/golang-jwt/jwt/pull/271) and the result in a simple case is basically what your original example does.

However, the problem is that in order for everything to work, I need to add type parameters to the `Parser` struct; because I need type parameters in `Parser.ParseWithClaims` (the global `ParseWithClaims` is just a wrapper basically around `NewParser`/`ParseWithClaims`) and we cannot have type parameters in methods (yet) (see https://github.com/golang/go/issues/49085).

This then leads to more ugliness, since I need to type parse `ParserOption` as well, leading to code like:
```go
token, err := jwt.ParseWithClaims(tokenString, &MyCustomClaims{}, func(token *jwt.Token[*MyCustomClaims]) (interface{}, error) {
   return []byte("AllYourBase"), nil
}, jwt.WithLeeway[*MyCustomClaims](5*time.Second))
```

which is really ugly because there is just no way to justify that I need to specify my claims type to the leeway parser option :(

**Reply by oxisto** [reply]
Update: I managed to remove the type parameter from `ParserOptions`, still a lot of stuck breaks :D

**Reply by costela** [reply]
I have a tentative version here: https://github.com/golang-jwt/jwt/pull/272
I couldn't resist and did some "API trimming" as well, assuming a new version is allowed some API breakage.

WDYT?
